go/oasis-test-runner/scenario/e2e/runtime_dynamic.go

```diff
@@ -103,14 +100,101 @@ func (sc *runtimeDynamicImpl) Run(childEnv *env.Env) error { // nolint: gocyclo
+	kmTxPath := filepath.Join(childEnv.Dir(), "register_km_runtime.json")
+	if err := cli.Registry.GenerateRegisterRuntimeTx(nonce, kmRtDesc, kmTxPath, ""); err != nil {
+		return fmt.Errorf("failed to generate register KM runtime tx: %w", err)
+	}
+	nonce++
+	if err := cli.Consensus.SubmitTx(kmTxPath); err != nil {
+		return fmt.Errorf("failed to register KM runtime: %w", err)
+	}
+
+	// Generate and update the new keymanager runtime's policy.
+	kmPolicyPath := filepath.Join(childEnv.Dir(), "km_policy.cbor")
+	kmPolicySig1Path := filepath.Join(childEnv.Dir(), "km_policy_sig1.pem")
+	kmPolicySig2Path := filepath.Join(childEnv.Dir(), "km_policy_sig2.pem")
+	kmPolicySig3Path := filepath.Join(childEnv.Dir(), "km_policy_sig3.pem")
+	kmUpdateTxPath := filepath.Join(childEnv.Dir(), "km_gen_update.json")
+	sc.logger.Info("building KM SGX policy enclave policies map")
+	enclavePolicies := make(map[sgx.EnclaveIdentity]*keymanager.EnclavePolicySGX)
+	kmRtEncID := kmRt.GetEnclaveIdentity()
+	var havePolicy bool
+	if kmRtEncID != nil {
+		enclavePolicies[*kmRtEncID] = &keymanager.EnclavePolicySGX{}
+		enclavePolicies[*kmRtEncID].MayQuery = make(map[common.Namespace][]sgx.EnclaveIdentity)
+		enclavePolicies[*kmRtEncID].MayReplicate = []sgx.EnclaveIdentity{}
+		for _, rt := range sc.net.Runtimes() {
+			if rt.Kind() != registry.KindCompute {
+				continue
+			}
+			if eid := rt.GetEnclaveIdentity(); eid != nil {
+				enclavePolicies[*kmRtEncID].MayQuery[rt.ID()] = []sgx.EnclaveIdentity{*eid}
+				// This is set only in SGX mode.
+				havePolicy = true
+			}
+		}
+	}
+	sc.logger.Info("initing KM policy")
+	if err := cli.Keymanager.InitPolicy(kmRt.ID(), 1, enclavePolicies, kmPolicyPath); err != nil {
+		return err
+	}
+	sc.logger.Info("signing KM policy")
+	if err := cli.Keymanager.SignPolicy("1", kmPolicyPath, kmPolicySig1Path); err != nil {
+		return err
+	}
+	if err := cli.Keymanager.SignPolicy("2", kmPolicyPath, kmPolicySig2Path); err != nil {
+		return err
+	}
+	if err := cli.Keymanager.SignPolicy("3", kmPolicyPath, kmPolicySig3Path); err != nil {
+		return err
+	}
+	if havePolicy {
+		// In SGX mode, we can update the policy as intended.
+		sc.logger.Info("updating KM policy")
+		if err := cli.Keymanager.GenUpdate(nonce, kmPolicyPath, []string{kmPolicySig1Path, kmPolicySig2Path, kmPolicySig3Path}, kmUpdateTxPath); err != nil {
+			return err
+		}
+		nonce++
+		if err := cli.Consensus.SubmitTx(kmUpdateTxPath); err != nil {
+			return fmt.Errorf("failed to update KM policy: %w", err)
+		}
+	} else {
+		// In non-SGX mode, the policy update fails with a policy checksum
+		// mismatch (the non-SGX KM returns an empty policy), so we need to
+		// do an epoch transition instead (to complete the KM runtime
+		// registration).
+		if err := sc.epochTransition(ctx); err != nil {
+			return err
+		}
+	}
+
+	// Wait for enough nodes to register, then make another epoch transition.
+	if havePolicy {
+		numNodes += len(sc.net.Keymanagers())
```

Comment:
Hm, I think both cases (SGX and non-SGX) should wait for the key manager to register before performing an epoch transition.

Reply:
Thanks, fixed! :)